### ISSUE TITLE
Tests fail when using Solidity custom errors & `configureYulOptimizer=true`

### ISSUE THREAD

**Issue by eladmallel** [sc-forks/solidity-coverage#717]
In contract code we're using custom errors, e.g.: 
```
error AdminOnly();

...

if (msg.sender != admin) {
    revert AdminOnly();
}
```

and then in the test code we have assertions like this:
```
await expect(
        gov.connect(account0)._setDynamicQuorumParams({
          minQuorumVotesBPS: 0,
          maxQuorumVotesBPS: 0,
          quorumVotesBPSOffset: 0,
          quorumLinearCoefficient: 0,
          quorumQuadraticCoefficient: 0,
        }),
      ).to.be.revertedWith('AdminOnly()');
```

When running tests under coverage this tests fails with the message:
```
AssertionError: Expected transaction to be reverted with AdminOnly(), but other exception was thrown: Error: VM Exception while processing transaction: reverted with an unrecognized custom error
```

Has anyone else run into this issue?
Any known solutions?

Thanks!

**Comment by cgewecke** [comment]
Could you provide a reproduction case or link to a repo where it's possible to investigate this?

**Comment by eladmallel** [comment]
Here: https://github.com/nounsDAO/nouns-monorepo/tree/verbs-nip-1-save-proposal-creation-block

Go into the contracts package and run:

`yarn coverage --testfiles test/governance/NounsDAO/V2/quorumConfig.test.ts`

**Comment by davidbrai** [comment]
This is the specific commit version where it fails: https://github.com/nounsDAO/nouns-monorepo/tree/11dc46ccf886767298934511db674350131ebf99

**Comment by cgewecke** [comment]
@davidbrai @eladmallel I think this is either a bug in Solidity or Hardhat but it will need further investigation. (If I can make a minimal reproduction case that's not coverage related I'll open an issue in the relevant repository.)

Things are as follows:
+ your contracts throw a stack too deep error when compiled without the optimizer on
+ solidity-coverage needs to turn the optimizer off (because it removes our tracing instrumentation)
+ there a [special optimizer config][1] that solves this conflict 
+ the custom error reporting doesn't work correctly with the special config

I believe this is the problem because if I turn the optimizer on in the plugin your test passes.

[1]: https://github.com/sc-forks/solidity-coverage/blob/bc7d07623ec4bd71a2ce4c8ecb4a872af613c8ef/plugins/hardhat.plugin.js#L62-L69

The only workaround I see in the near term is to detect that solidity-coverage is running and make your test less specific under those conditions, something like:

```js
it.only('reverts when sender is not admin', async () => {
  // solidity-coverage automatically sets this variable on the env so you can deal with its weird problems
  if (hre.__SOLIDITY_COVERAGE_RUNNING === true) {
    await expect(
      gov.connect(account0)._setDynamicQuorumParams({
        minQuorumVotesBPS: 0,
        maxQuorumVotesBPS: 0,
        quorumVotesBPSOffset: 0,
        quorumLinearCoefficient: 0,
        quorumQuadraticCoefficient: 0,
      }),
    ).to.be.reverted();
  } else {
   // normal test  which checks for 'AdminOnly'
  }
});

**Comment by eladmallel** [comment]
Thank you @cgewecke! For now we're decided to skip these tests during the coverage run using the mocha grep configuration.
We appreciate your support and your work on this awesome plugin 🙏

**Comment by cgewecke** [comment]
Possibly related issues:
+ [hh 2567][1]
+ [hh 2546][2]


[1]: https://github.com/NomicFoundation/hardhat/issues/2657
[2]: https://github.com/NomicFoundation/hardhat/issues/2546

**Comment by cgewecke** [comment]
This has been fixed at Hardhat in 2.10.1

https://github.com/NomicFoundation/hardhat/issues/2546#issuecomment-1183754292
